src/fireEvent.ts

```diff
@@ -8,8 +9,21 @@ const isHostElement = (element?: ReactTestInstance) => {
 };
 
 const isTextInput = (element?: ReactTestInstance) => {
+  if (!element) {
+    return false;
+  }
+
   const { TextInput } = require('react-native');
-  return element?.type === TextInput;
+  // We have to test if the element type is either the TextInput component
+  // (which would if it is a composite component) or the string
+  // TextInput (which would be true if it is a host component)
+  // All queries but the one by testID return composite component and event
+  // if all queries returned host components, since fireEvent bubbles up
+  // it would trigger the parent prop without the composite component check
+  return (
+    filterNodeByType(element, TextInput) ||
```

Comment:
Wondering if this should be handled directly in `filterNodeByType`? That'd require a bit more refactoring, and some form of `host component -> composite` element mapping. It would also add some overhead in branches that don't really care about it (all the `Text` based one for instance since we're not using `byID`.

So fine with me, but perhaps some other maintainer might feel different about that.

Reply:
You mean if filterNodeByType made the comparison for both the component and the component name ? It would involve that every query would need to filter composite components afterwards (except testId queries that already filter host components), I think it would introduce additional complexity and I don't see clearly how it could be beneficial so I'm not sure about it either

Reply:
> You mean if filterNodeByType made the comparison for both the component and the component name ? It would involve that every query would need to filter composite components afterwards (except testId queries that already filter host components), I think it would introduce additional complexity and I don't see clearly how it could be beneficial so I'm not sure about it either

Yeah exactly. I was wondering if that would be the sort of abstraction we would like to move out of `fireEvent`. But given it's a very specific usecase for now, and given the complexity it implies, I agree that it's not probably worth it (hence my approval)

Reply:
BTW I am against having `filterNodeByType` to match against both `type` and `type.name` has it would get confusing what thing will you actually get back.

Reply:
Yes TextInput (composite) is always rendering a host TextInput, I was suggesting events could only be triggered on host component in the discussion on the other pr https://github.com/callstack/react-native-testing-library/pull/1080#issuecomment-1237429774 but as @AugustinLF noticed, it would be an issue for props with different naming. This could maybe be fixed by mapping native props name with composite components props name but it seems rather complicated and I'm not sure such a mapping is entirely possible. Also props from custom components that do not match any native prop couldn't be triggered through fake events anymore

Reply:
re query returning host components: iirc we had this assumption with @thymikee that all of the queries should always return host components (maybe except the `UNSAFE_getBy...` queries), just like in RTL where you are given actual DOM components (like `div`, etc), and not composite components as only host components result directly with "visible" UI.

re `TextInput` predicate: we have have at least two options for now
1) Find out which host TextInput prop is set by composite TextInput component when it has `editable=false`.
2) Assuming that component structure is fixed we might reach from host TextInput to parent composite TextInput using `node.parent` to verify its props.

Both of the methods rely in some way on "internal" component structure from React Native, the question is which is less likely to change. I suspect that 1st option might be more stable as it probably refers to touch event handling, but we would need to research which prop is actually being set when `editable=false`.

@pierrezimmermannbam, @AugustinLF, @thymikee which one would you pick, do you see other alternatives?

Additionally we should create a test case where we test whether our simulation works fine, so that we are able to detect when things break. That probably should be something like:

```ts
test(`fireEvent handles implicitly editable TextInput', () => {
  const onChange = jest.fn();
  render(<TextInput testID="subject" onChange={jest.fn} />);
  const subject = screen.getByTest("subject");
  expect(subject.type).toBe("TextInput");
  fireEvent.change(subject, "new value");
  expect(onChange).toHaveBeenCalledWith("newValue");
}

test(`fireEvent handles explicitly editable TextInput', () => {
  const onChange = jest.fn();
  render(<TextInput testID="subject" onChange={jest.fn} editable />);
  const subject = screen.getByTest("subject");
  expect(subject.type).toBe("TextInput");
  fireEvent.change(subject, "new value");
  expect(onChange).toHaveBeenCalledWith("newValue");
}

test(`fireEvent handles non-editable TextInput', () => {
  const onChange = jest.fn();
  render(<TextInput testID="subject" onChange={jest.fn} editable={false} />);
  const subject = screen.getByTest("subject");
  expect(subject.type).toBe("TextInput");
  fireEvent.change(subject, "new value");
  expect(onChange).not.toHaveBeenCalled();
}
```